File: packages/core/src/core/text-constraints.ts
/**
 * Declarative text-constraint builders — `requires` / `forbids` (section-scoped) and
 * `textRule` (document-scoped). They attach required / forbidden phrase checks to a contract,
 * compiling to the engine's existing node-local `rule` / cross-plane `docRule` machinery
 * (D-0011 / C-0009): "does this literal-or-regex appear (or not) in this scope, the right
 * number of times?". This is the TS-API surface the declarative `requires:` / `forbids:` YAML
 * keys compile to, so both authoring surfaces share one implementation and one set of findings.
 *
 * Each builder is a thin scope-resolver over the text-match core (`text-match.ts`):
 *
 *   - `requires(specs)` / `forbids(specs)` → a node-local `Rule`. The matcher runs over the
 *     bound section's *subtree text* (every block — prose, list items, table cells, code — of
 *     the section and its descendants, reconstructed line-faithfully from the projected
 *     `SectionNode` so a `forbids` hit reports its real source line). The `scopeKey` is the
 *     section's generated lowerCamelCase key; a `requires` miss / count violation pins at the
 *     section heading.
 *   - `textRule({ requires?, forbids? })` → a cross-plane `DocRule` over the WHOLE document.
 *     The `DocRule` now also receives the projected `DocTree`, so the whole-document text is
 *     reconstructed line-faithfully from the projected tree (`tree.root`) — a whole-document
 *     `forbids` hit therefore anchors at the exact offending source line. The `scopeKey` is the
 *     literal `"doc"`; a `requires` miss is document-level (no position).
 *
 * Purity (D-0011): a `requires` entry whose bound expresses ABSENCE (`max: 0`, or `max < min`)
 * is a constructor-time error — the absence form is `forbids`. The check fires when `requires(...)`
 * (or the `requires` arm of `textRule(...)`) is *called*, not at validation time.
 */
import { toCamelKey } from "./camel.js";
import { ContractBuildError } from "./grammar.js";
import { buildTextFindings, matchText } from "./text-match.js";
import type { TextMatchSpec } from "./text-match.js";
import type { BlockNode, Ctx, Doc, DocRule, DocTree, Finding, Rule, SectionNode } from "./types.js";

/**
 * One required / forbidden text-match entry — the closed match-spec vocabulary (D-0011 § The
 * match spec). This is the matcher's `TextMatchSpec` re-exported verbatim, so the authoring
 * surface and the matcher core never diverge: exactly one of `pattern` (a literal substring) or
 * `regex` (a regex source) supplies the needle; `normalize` / `ignoreCase` tune matching;
 * `min` / `max` bound the count; `id` pins identity; `note` / `level` shape the finding.
 */
export type { TextMatchSpec } from "./text-match.js";

/**
 * The document-scoped options for `textRule(...)` — `requires` / `forbids` lists, each a set of
 * independent whole-document checks (the body-root form of D-0011's `requires:` / `forbids:`).
 */
export interface TextRuleSpec {
  requires?: TextMatchSpec[];
  forbids?: TextMatchSpec[];
}

// ── Purity (D-0011): a `requires` entry may not express absence ────────────────────────

/** The needle as it reads in an error message — `"literal"` for a pattern, `/source/` for a regex. */
function specRepr(spec: TextMatchSpec): string {
  if (spec.regex !== undefined) return `/${spec.regex}/`;
  return `"${spec.pattern ?? ""}"`;
}

/**
 * Reject a `requires` entry whose count bound expresses ABSENCE — `max: 0`, or any `max` below
 * the entry's effective minimum (`min`, default 1). A "this scope must NOT contain X" check is
 * the `forbids` form, not a `requires` with `max: 0`. Thrown at construction (a contract-authoring
 * error), mirroring `grammar.ts`'s build-time `ContractBuildError` convention.
 */
function assertRequiresPurity(specs: TextMatchSpec[]): void {
  for (const spec of specs) {
    const floor = Math.max(spec.min ?? 1, 1);
    if (spec.max !== undefined && spec.max < floor) {
      throw new ContractBuildError(
        "contract/text-requires-purity",
        `requires(${specRepr(spec)}): a requires entry may not express absence ` +
          `(max ${spec.max} < ${floor}); use forbids(...) for an absence check`,
      );
    }
  }
}

// ── Line-faithful scope-text reconstruction ────────────────────────────────────────────

/**
 * Accumulates text fragments at their 1-based source lines and renders them into one string
 * whose newline structure mirrors the source — so the matcher's positions land on real source
 * lines. A fragment may itself be multi-line (a paragraph's soft-wrapped prose); each of its
 * lines is placed on a consecutive source line. Two fragments on one line are space-joined.
 */
class LineBuffer {
  private lines: string[] = [];

  place(line: number, text: string): void {
    if (text === "") return;
    const parts = text.split("\n");
    for (let i = 0; i < parts.length; i++) {
      const idx = line - 1 + i;
      if (idx < 0) continue;
      while (this.lines.length <= idx) this.lines.push("");
      const existing = this.lines[idx]!;
      this.lines[idx] = existing === "" ? parts[i]! : `${existing} ${parts[i]!}`;
    }
  }

  render(): string {
    return this.lines.join("\n");
  }
}

/** Place one projected block's text at its real source line(s). */
function placeBlock(buf: LineBuffer, block: BlockNode): void {
  switch (block.kind) {
    case "paragraph":
      buf.place(block.pos.line, block.text);
      break;
    case "list":
      for (const item of block.items) buf.place(item.pos.line, item.text);
      break;
    case "code":
      buf.place(block.pos.line, block.value);
      break;
    case "table":
      buf.place(block.pos.line, block.columns.join(" "));
      block.rows.forEach((row, i) => buf.place(block.rowPos(i).line, row.join(" ")));
      break;
  }
}

/**
 * The section's subtree text, reconstructed line-faithfully from the projected `SectionNode`:
 * every block (prose, list items, table cells, code) of this section and every descendant,
 * placed at its real source line. Built from the `SectionNode` (not the typed `SectionView`),
 * so per-block positions are exact and a `forbids` hit reports its true source line.
 */
function sectionScopeText(node: SectionNode): string {
  const buf = new LineBuffer();
  const walk = (n: SectionNode): void => {
    for (const block of n.blocks) placeBlock(buf, block);
    for (const child of n.sections) walk(child);
  };
  walk(node);
  return buf.render();
}

/** The whole-document scope text, reconstructed line-faithfully from the projected tree. */
function docScopeText(tree: DocTree): string {
  return sectionScopeText(tree.root);
}

// ── The builders ───────────────────────────────────────────────────────────────────────

/**
 * Require each listed phrase to be PRESENT in the bound section's subtree text — a node-local
 * `Rule` for a section's `rules: [...]` slot (the section-scoped form of D-0011's `requires:`).
 * Each entry emits its own `text/requires` finding at the section heading when its phrase is
 * absent (or `text/count` when a `min` / `max` bound is violated). Rejects an absence-form
 * entry (`max: 0` / `max < min`) at construction (use {@link forbids}).
 */
export function requires(specs: TextMatchSpec[]): Rule {
  assertRequiresPurity(specs);
  return {
    __brand: "Rule",
    id: "text/requires",
    run(node: SectionNode, ctx: Ctx): Finding[] {
      const scopeKey = toCamelKey(node.name);
      const text = sectionScopeText(node);
      const out: Finding[] = [];
      for (const spec of specs) {
        out.push(
          ...buildTextFindings({
            kind: "requires",
            spec,
            match: matchText(text, spec),
            scopeKey,
            scope: node.name,
            scopePos: node.pos,
            ctx,
          }),
        );
      }
      return out;
    },
  };
}

/**
 * Forbid each listed phrase from appearing in the bound section's subtree text — a node-local
 * `Rule` for a section's `rules: [...]` slot (the section-scoped form of D-0011's `forbids:`).
 * Each entry emits a `text/forbids` finding at the offending line for every hit (or `text/count`
 * at the heading when a positive `max` cap is exceeded). `forbids` is the absence form, so there
 * is no purity restriction.
 */
export function forbids(specs: TextMatchSpec[]): Rule {
  return {
    __brand: "Rule",
    id: "text/forbids",
    run(node: SectionNode, ctx: Ctx): Finding[] {
      const scopeKey = toCamelKey(node.name);
      const text = sectionScopeText(node);
      const out: Finding[] = [];
      for (const spec of specs) {
        out.push(
          ...buildTextFindings({
            kind: "forbids",
            spec,
            match: matchText(text, spec),
            scopeKey,
            scope: node.name,
            scopePos: node.pos,
            ctx,
          }),
        );
      }
      return out;
    },
  };
}

/**
 * Attach required / forbidden phrase checks to the WHOLE document — a cross-plane `DocRule` for
 * a contract's `rules: [...]` slot (the body-root form of D-0011's `requires:` / `forbids:`).
 * Each `requires` entry emits a document-level `text/requires` (no position) when its phrase is
 * absent; each `forbids` entry emits a `text/forbids` at the offending line for every hit. The
 * `requires` arm enforces the same absence-form purity as {@link requires}.
 */
export function textRule(spec: TextRuleSpec): DocRule {
  const requiresSpecs = spec.requires ?? [];
  const forbidsSpecs = spec.forbids ?? [];
  assertRequiresPurity(requiresSpecs);
  return {
    __brand: "DocRule",
    id: "text/doc",
    run(doc: Doc, ctx: Ctx, tree: DocTree): Finding[] {
      const text = docScopeText(tree);
      const out: Finding[] = [];
      for (const s of requiresSpecs) {
        out.push(
          ...buildTextFindings({
            kind: "requires",
            spec: s,
            match: matchText(text, s),
            scopeKey: "doc",
            ctx,
          }),
        );
      }
      for (const s of forbidsSpecs) {
        out.push(
          ...buildTextFindings({
            kind: "forbids",
            spec: s,
            match: matchText(text, s),
            scopeKey: "doc",
            ctx,
          }),
        );
      }
      return out;
    },
  };
}

File: crates/markdown-contract-engine/src/text_constraints.rs
//! Declarative text-constraint builders — [`requires`] / [`forbids`] (section-scoped
//! node rules) and [`text_rule`] (a document-scoped rule), ported from the TS
//! `text-constraints.ts` (D-0011 / C-0009).
//!
//! Each builder is a thin scope-resolver over the text-match core
//! ([`crate::text_match`]): the matcher runs over the bound scope's text, reconstructed
//! **line-faithfully** from the projected tree — every block (prose, list items, table
//! cells, code) placed at its real source line — so a `forbids` hit reports its true
//! source line. A section scope's `scope_key` is the heading's generated
//! lowerCamelCase key; the document scope's is the literal `"doc"`.
//!
//! Purity (D-0011): a `requires` entry whose bound expresses ABSENCE (`max: 0`, or
//! `max` below the effective minimum) panics at construction — the absence form is
//! `forbids`. This mirrors the TS builders' `ContractBuildError` throw; the declarative
//! compiler screens the same shape out ahead of time as a `DeclarativeError`.

use crate::camel::to_camel_key;
use crate::contract::{DocRule, Rule};
use crate::finding::Finding;
use crate::registry::Ctx;
use crate::text_match::{
    TextFindingInput, TextKind, TextMatchSpec, build_text_findings, match_text,
};
use crate::tree::{BlockNode, DocTree, SectionNode};

// ── Purity (D-0011): a `requires` entry may not express absence ───────────────────────

/// The needle as it reads in an error message.
fn spec_repr(spec: &TextMatchSpec) -> String {
    if let Some(regex) = &spec.regex {
        format!("/{regex}/")
    } else {
        format!("\"{}\"", spec.pattern.as_deref().unwrap_or(""))
    }
}

/// Reject a `requires` entry whose count bound expresses ABSENCE (`max: 0`, or any
/// `max` below the entry's effective minimum). Panics at construction — a
/// contract-authoring error, mirroring the TS `ContractBuildError`.
fn assert_requires_purity(specs: &[TextMatchSpec]) {
    for spec in specs {
        let floor = spec.min.unwrap_or(1.0).max(1.0);
        if let Some(max) = spec.max
            && max < floor
        {
            panic!(
                "contract/text-requires-purity: requires({}): a requires entry may not express absence (max {max} < {floor}); use forbids(...) for an absence check",
                spec_repr(spec)
            );
        }
    }
}

// ── Line-faithful scope-text reconstruction ───────────────────────────────────────────

/// Accumulates text fragments at their 1-based source lines and renders them into one
/// string whose newline structure mirrors the source. A multi-line fragment places each
/// of its lines on a consecutive source line; two fragments on one line space-join.
struct LineBuffer {
    lines: Vec<String>,
}

impl LineBuffer {
    fn new() -> Self {
        Self { lines: Vec::new() }
    }

    fn place(&mut self, line: u32, text: &str) {
        if text.is_empty() {
            return;
        }
        for (i, part) in text.split('\n').enumerate() {
            let Some(idx) = (line as usize + i).checked_sub(1) else {
                continue;
            };
            while self.lines.len() <= idx {
                self.lines.push(String::new());
            }
            let existing = &mut self.lines[idx];
            if existing.is_empty() {
                existing.push_str(part);
            } else {
                existing.push(' ');
                existing.push_str(part);
            }
        }
    }

    fn render(&self) -> String {
        self.lines.join("\n")
    }
}

/// Place one projected block's text at its real source line(s).
fn place_block(buf: &mut LineBuffer, block: &BlockNode) {
    match block {
        BlockNode::Paragraph { text, pos, .. } => buf.place(pos.line, text),
        BlockNode::List { items, .. } => {
            for item in items {
                buf.place(item.pos.line, &item.text);
            }
        }
        BlockNode::Code { value, pos, .. } => buf.place(pos.line, value),
        BlockNode::Table {
            columns,
            rows,
            row_pos,
            pos,
            ..
        } => {
            buf.place(pos.line, &columns.join(" "));
            for (i, row) in rows.iter().enumerate() {
                if let Some(rp) = row_pos.get(i) {
                    buf.place(rp.line, &row.join(" "));
                }
            }
        }
    }
}

/// The section's subtree text, reconstructed line-faithfully: every block of this
/// section and every descendant, placed at its real source line.
fn section_scope_text(node: &SectionNode) -> String {
    fn walk(n: &SectionNode, buf: &mut LineBuffer) {
        for block in &n.blocks {
            place_block(buf, block);
        }
        for child in &n.sections {
            walk(child, buf);
        }
    }
    let mut buf = LineBuffer::new();
    walk(node, &mut buf);
    buf.render()
}

/// The whole-document scope text, reconstructed line-faithfully from the projected tree.
fn doc_scope_text(tree: &DocTree) -> String {
    section_scope_text(&tree.root)
}

// ── The builders ──────────────────────────────────────────────────────────────────────

struct TextNodeRule {
    kind: TextKind,
    id: &'static str,
    specs: Vec<TextMatchSpec>,
}

impl Rule for TextNodeRule {
    fn id(&self) -> &str {
        self.id
    }

    fn run(&self, node: &SectionNode, ctx: &Ctx) -> Vec<Finding> {
        let scope_key = to_camel_key(&node.name);
        let text = section_scope_text(node);
        let mut out = Vec::new();
        for spec in &self.specs {
            out.extend(build_text_findings(
                TextFindingInput {
                    kind: self.kind,
                    spec,
                    result: match_text(&text, spec),
                    scope_key: &scope_key,
                    scope: Some(&node.name),
                    scope_pos: Some(node.pos),
                },
                ctx,
            ));
        }
        out
    }
}

/// Require each listed phrase to be PRESENT in the bound section's subtree text — a
/// node-local [`Rule`] for a section's `rules` slot. Each entry emits its own
/// `text/requires` finding at the section heading when its phrase is absent (or
/// `text/count` when a `min` / `max` bound is violated).
///
/// # Panics
/// On an absence-form entry (`max: 0` / `max < min`) — use [`forbids`].
pub fn requires(specs: Vec<TextMatchSpec>) -> Box<dyn Rule> {
    assert_requires_purity(&specs);
    Box::new(TextNodeRule {
        kind: TextKind::Requires,
        id: "text/requires",
        specs,
    })
}

/// Forbid each listed phrase from appearing in the bound section's subtree text — a
/// node-local [`Rule`]. Each entry emits a `text/forbids` finding at the offending line
/// for every hit (or `text/count` at the heading when a positive `max` cap is exceeded).
pub fn forbids(specs: Vec<TextMatchSpec>) -> Box<dyn Rule> {
    Box::new(TextNodeRule {
        kind: TextKind::Forbids,
        id: "text/forbids",
        specs,
    })
}

/// The document-scoped options for [`text_rule`] — `requires` / `forbids` lists, each a
/// set of independent whole-document checks.
#[derive(Default)]
pub struct TextRuleSpec {
    pub requires: Vec<TextMatchSpec>,
    pub forbids: Vec<TextMatchSpec>,
}

struct TextDocRule {
    spec: TextRuleSpec,
}

impl DocRule for TextDocRule {
    fn id(&self) -> &str {
        "text/doc"
    }

    fn run(&self, tree: &DocTree, ctx: &Ctx) -> Vec<Finding> {
        let text = doc_scope_text(tree);
        let mut out = Vec::new();
        for spec in &self.spec.requires {
            out.extend(build_text_findings(
                TextFindingInput {
                    kind: TextKind::Requires,
                    spec,
                    result: match_text(&text, spec),
                    scope_key: "doc",
                    scope: None,
                    scope_pos: None,
                },
                ctx,
            ));
        }
        for spec in &self.spec.forbids {
            out.extend(build_text_findings(
                TextFindingInput {
                    kind: TextKind::Forbids,
                    spec,
                    result: match_text(&text, spec),
                    scope_key: "doc",
                    scope: None,
                    scope_pos: None,
                },
                ctx,
            ));
        }
        out
    }
}

/// Attach required / forbidden phrase checks to the WHOLE document — a cross-plane
/// [`DocRule`] for a contract's `rules` slot. A `requires` miss is document-level (no
/// position); each `forbids` hit anchors at the exact offending source line.
///
/// # Panics
/// On an absence-form `requires` entry (same purity as [`requires`]).
pub fn text_rule(spec: TextRuleSpec) -> Box<dyn DocRule> {
    assert_requires_purity(&spec.requires);
    Box::new(TextDocRule { spec })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::contract::{Contract, LevelOpts, SectionOpts, section, section_with, sections};
    use crate::finding::FindingLevel;
    use crate::validate::validate;

    fn pattern(p: &str) -> TextMatchSpec {
        TextMatchSpec {
            pattern: Some(p.into()),
            ..Default::default()
        }
    }

    fn brief(findings: &[Finding]) -> Vec<(&str, FindingLevel, Option<u32>)> {
        findings
            .iter()
            .map(|f| (f.id.as_str(), f.level, f.pos.map(|p| p.line)))
            .collect()
    }

    // Contract first — the v22 fixture shape: a section-scoped requires misses and pins
    // the section heading with the synthesized id the goldens pin.
    #[test]
    fn section_requires_miss_pins_the_heading() {
        let build = || {
            Contract::new().body(sections(
                LevelOpts::default(),
                vec![section_with(
                    "Summary",
                    SectionOpts {
                        rules: vec![requires(vec![pattern("outcome")])],
                        ..Default::default()
                    },
                )],
            ))
        };
        let pass = "## Summary\n\nThe outcome is recorded here.\n";
        assert_eq!(validate(pass, &build(), "d.md"), vec![]);
        let fail = "## Summary\n\nThis decision records the chosen direction.\n";
        assert_eq!(
            brief(&validate(fail, &build(), "d.md")),
            vec![(
                "text/requires/summary/1tc7itx",
                FindingLevel::Error,
                Some(1)
            )]
        );
    }

    // The v23 shape: a document-scoped forbids hit anchors at the offending line.
    #[test]
    fn doc_forbids_hit_anchors_the_offending_line() {
        let build = || {
            Contract::new()
                .body(sections(LevelOpts::default(), vec![section("Summary")]))
                .doc_rule(text_rule(TextRuleSpec {
                    forbids: vec![TextMatchSpec {
                        normalize: Some(false),
                        ..pattern("}scripts/")
                    }],
                    ..Default::default()
                }))
        };
        let fail = "## Summary\n\nCallers reach into }scripts/legacy.sh directly.\n";
        assert_eq!(
            brief(&validate(fail, &build(), "d.md")),
            vec![("text/forbids/doc/o9pijh", FindingLevel::Error, Some(3))]
        );
    }

    // Scope text is the SUBTREE: a child section's list items are in scope, at their
    // real source lines.
    #[test]
    fn scope_text_reconstruction_is_line_faithful() {
        let tree = crate::parse::parse_document(
            "## Top\n\nintro prose\n\n### Child\n\n- item one\n- item two\n\n| A | B |\n| - | - |\n| x | y |\n",
        );
        let text = section_scope_text(&tree.root.sections[0]);
        let lines: Vec<&str> = text.split('\n').collect();
        assert_eq!(lines[2], "intro prose"); // line 3
        assert_eq!(lines[6], "item one"); // line 7
        assert_eq!(lines[9], "A B"); // line 10 (table header)
        assert_eq!(lines[11], "x y"); // line 12 (data row; separator line stays empty)
    }

    #[test]
    #[should_panic(expected = "text-requires-purity")]
    fn requires_rejects_the_absence_form() {
        requires(vec![TextMatchSpec {
            max: Some(0.0),
            ..pattern("x")
        }]);
    }
}
